error_reporting.py

```diff
@@ -0,0 +1,479 @@
+            # This is our error node
+            reason = determine_reason(node)
+            print_reason(reason)
+
+            # used_node_name is optional because could not determine the reason
+            used_node_name = extract_and_print_why(
+                func, reason, node, node_name)
+            eprint("FAILING ASSERTION", style="red on white", justify="center")
+            node_as_ap = node_dsa_to_node_ap(node)
+            eprint("ASSERT", pretty_node(node_as_ap))
+            if used_node_name is not None:
+                used_node = func.nodes[used_node_name]
+                used_node_as_ap = node_dsa_to_node_ap(used_node)
+
+                eprint("HINT: SUSPECTED STATEMENT",
+                       justify="center", style="red on white")
+                eprint(pretty_node(used_node_as_ap), style="magenta bold")
+
+            succ_smtlib_with_model = smt.make_smtlib(
+                prog, not_taken_path.union(set(successors)), with_model=True)
+            succ_model = send_smtlib_model(
+                succ_smtlib_with_model, smt.SolverZ3())
+
+            node_vars = set(
+                map(ap.convert_expr_var, source.used_variables_in_node(node)))
+
+            # need to get AP name
+            eprint("VARIABLES USED IN ASSERTION",
+                   justify="center", style="red on white")
+            get_relevant_responses(node_vars, succ_model)
+
+            return (reason, node_name, used_node_name)
+
+        # When len(successors) == 1 and it is a NodeCond, it is because the succ_else path to NodeNameErr was trimmed
+        # This is handled above, so we skip it.
+        if isinstance(node, source.NodeCond) and len(successors) != 1:
+            node1 = successors[0]
+            node2 = successors[1]
+            not_taken_path_and_succ1 = not_taken_path.union(set([node1]))
+            not_taken_path_and_succ2 = not_taken_path.union(set([node2]))
+            succ_node1_smtlib = smt.make_smtlib(prog, not_taken_path_and_succ1)
+            succ_node2_smtlib = smt.make_smtlib(prog, not_taken_path_and_succ2)
+            succ_node1_consistent, succ_node1_sat = get_sat(succ_node1_smtlib)
+            succ_node2_consistent, succ_node2_sat = get_sat(succ_node2_smtlib)
+            if succ_node1_sat == smt.CheckSatResult.UNSAT and succ_node2_sat == smt.CheckSatResult.UNSAT:
+                if succ_node1_consistent:
+                    q = q.union(set([node1]))
+                elif succ_node2_consistent:
+                    q = q.union(set([node2]))
+                else:
+                    assert False, "one path has to be consistent"
+                # EDGE case take any path but do not add to not_taken_path
+                q = q.union(set([node1]))
+            elif succ_node1_sat == smt.CheckSatResult.UNSAT and succ_node2_sat == smt.CheckSatResult.SAT:
+                q.add(node1)
+                not_taken_path.add(node2)
```

Comment:
This doesn't make sense to me, I would've expected it to be the other way around. According to my understanding:

```
if left is UNSAT and right SAT:
    # it means that, when assuming the left node is ok,
    # (ie. can't reach the error node going through it)
    # the program is correct (ie. can't reach the error node
    # from start node)
    # so the error must come from the right path

    q.add(right) # keep exploring the path where the error comes from

    not_taken_path.add(left)
```

But we do the opposite here.